onnxruntime/core/optimizer/fuse_initializers_transformer.cc

```diff
@@ -0,0 +1,172 @@
+ *
+ * @param graph            Graph object.
+ * @param node             Node object.
+ * @param tensor_type      The type of initialized tensor to be found in the given node.
+ * @param output_type      The type of output for the given node.
+ *
+ * @return  True, if for the given node an initialized tensor of "tensor_type" with specific "output_type" is found,
+ *          else, False.
+ */
+static bool IsNodeValidForFusion(const Graph& graph,
+                                 const Node& node,
+                                 const onnxruntime::MLDataType tensor_type,
+                                 const onnxruntime::MLDataType output_type) {
+  // Node must have initialized tensor
+  if (!(graph.IsInitializedTensor(node.InputDefs()[0]->Name()))) return false;
+
+  // Initialzed tensor must be of tensor_type
+  if (!(DataTypeImpl::TypeFromProto(*(node.InputDefs()[0]->TypeAsProto())) == tensor_type)) return false;
+
+  // Node output must be of output_type
+  if (!(DataTypeImpl::TypeFromProto(*(node.OutputDefs()[0]->TypeAsProto())) == output_type)) return false;
+
+  return true;
+}
+
+/**
+ * @brief   Make a new name from the old node arg name.
+ *
+ * It replaces "InsertedPrecisionFreeCast_" prefix in a node name with "FusedBack_" prefix.
+ *
+ * @param   old_node_arg_name Old arg name.
+ *
+ * @return  New arg name.
+ */
+static const std::string NewNodeArgName(const std::string& old_node_arg_name) {
+  static thread_local const std::string pattern_to_be_replaced = "InsertedPrecisionFreeCast_";
+  std::string new_node_arg_name = old_node_arg_name;
+  auto pos = new_node_arg_name.find(pattern_to_be_replaced);
+  if (std::string::npos != pos) new_node_arg_name.replace(pos, pattern_to_be_replaced.size(), "");
+  new_node_arg_name = "FusedBack_" + new_node_arg_name;
+  return new_node_arg_name;
+}
+
+/**
+ * @brief   It fuses the initializer in the current node to its next node / output node, and then
+ *          remove the link/edge between current node and next node / output node.
+ *
+ * The node input_defs_index and output_defs_index is assumed to be always 0, as the node which encapsulates
+ * a single Initializer have just zero input, one initializer and one output.
+ *
+ * @param graph                 Graph object.
+ * @param node                  Current Node to be fused with its next node.
+ * @param next_node_arg_type    The "type" of initializer expected by the next-node to the current node.
+ * @param thread_pool           Thread pool for multi-threaded conversion of the initializer
+ *                              from an unsupported to supported type tensor.
+ */
+static void FuseInitializerWithNode(Graph& graph,
+                                    Node& node,
+                                    const onnxruntime::MLDataType next_node_arg_type,
+                                    onnxruntime::concurrency::ThreadPool* thread_pool) {
+  // Get next node
+  Node& next_node = *graph.GetNode(node.OutputNodesBegin()->Index());
```

Comment:
There might not be next node: graph.GetNode(node.OutputNodesBegin()->Index()) could be nullptr.